File: Android/app/src/main/java/com/ln/collectionview/scrollview/pulse/LNScrollViewPulseConvertor.kt
package com.ln.collectionview.scrollview.pulse

import java.lang.ref.WeakReference

class LNScrollViewPulseConvertor: LNScrollViewPulseGeneratorDelegate {

    private var generator: LNScrollViewPulseGenerator? = null
    private var pulser: LNScrollViewPulser? = null
    var isConversationOfEnergy: Boolean = false

    init {
        isConversationOfEnergy = false
    }

    fun bindGenerator(generator: LNScrollViewPulseGenerator) {
        this.generator?.delegate = null
        this.generator = generator
        generator.delegate = WeakReference(this)
    }

    fun bindPulser(pulser: LNScrollViewPulser) {
        this.pulser = pulser
    }

    override fun generatorHasDetectedMomentum(momentum: LNScrollViewMomentum): LNScrollViewMomentum {
        return if (isConversationOfEnergy) {
            val pulserMomentum = pulser?.getCurrentMomentum() ?: return LNScrollViewMomentum().apply {
                mass = momentum.mass
                velocity = 0f
            }
            val targetMomentum = LNScrollViewMomentum().apply {
                mass = pulserMomentum.mass
                velocity = (2 * momentum.mass * momentum.velocity + pulserMomentum.mass * pulserMomentum.velocity - momentum.mass * pulserMomentum.velocity) /
                        (momentum.mass + pulserMomentum.mass)
            }
            pulser?.updateMomentum(targetMomentum)

            val feedbackMomentum = LNScrollViewMomentum().apply {
                mass = momentum.mass
                velocity = (2 * pulserMomentum.mass * pulserMomentum.velocity + momentum.mass * momentum.velocity - pulserMomentum.mass * momentum.velocity) /
                        (momentum.mass + pulserMomentum.mass)
            }
            feedbackMomentum
        } else {
            val pulserMomentum = pulser?.getCurrentMomentum() ?: return LNScrollViewMomentum().apply {
                mass = momentum.mass
                velocity = 0f
            }
            val targetMomentum = LNScrollViewMomentum().apply {
                mass = pulserMomentum.mass
                velocity = (momentum.mass * momentum.velocity + pulserMomentum.mass * pulserMomentum.velocity) / pulserMomentum.mass
            }
            pulser?.updateMomentum(targetMomentum)

            val feedbackMomentum = LNScrollViewMomentum().apply {
                mass = momentum.mass
                velocity = 0f
            }
            feedbackMomentum
        }
    }
}

File: iOS/LNCollectionView/LNScrollView/Pulse/LNScrollViewPulseConvertor.m
//
//  LNScrollViewPulseConvertor.m
//  LNCollectionView
//
//  Created by Levison on 28.11.24.
//

#import "LNScrollViewPulseConvertor.h"

@interface LNScrollViewPulseConvertor()

@property (nonatomic, weak) LNScrollViewPulseGenerator *generator;
@property (nonatomic, weak) LNScrollViewPulser *pulser;

@end

@implementation LNScrollViewPulseConvertor

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.isConversationOfEnergy = NO;
    }
    return self;
}

- (void)bindGenerator:(LNScrollViewPulseGenerator *)generator
{
    if (self.generator && self.generator.delegate == self) {
        self.generator.delegate = nil;
    }
    _generator = generator;
    generator.delegate = self;
}

- (void)bindPulser:(LNScrollViewPulser *)pulser
{
    self.pulser = pulser;
}

- (LNScrollViewMomentum *)generatorHasDetectedMomentum:(LNScrollViewMomentum *)momentum
{
    if (self.isConversationOfEnergy) {
        LNScrollViewMomentum *pulserMomentum = [self.pulser getCurrentMomentum];
        LNScrollViewMomentum *targetMomentum = [[LNScrollViewMomentum alloc] init];
        targetMomentum.mass = pulserMomentum.mass;
        targetMomentum.velocity = (2 * momentum.mass * momentum.velocity + pulserMomentum.mass * pulserMomentum.velocity - momentum.mass * pulserMomentum.velocity)/(momentum.mass + pulserMomentum.mass);
        [self.pulser updateMomentum:targetMomentum];
        LNScrollViewMomentum *feedbackMomentum = [[LNScrollViewMomentum alloc] init];
        feedbackMomentum.mass = momentum.mass;
        feedbackMomentum.velocity = (2 * pulserMomentum.mass * pulserMomentum.velocity + momentum.mass * momentum.velocity - pulserMomentum.mass * momentum.velocity)/(momentum.mass + pulserMomentum.mass);
        return feedbackMomentum;
    } else {
        LNScrollViewMomentum *pulserMomentum = [self.pulser getCurrentMomentum];
        LNScrollViewMomentum *targetMomentum = [[LNScrollViewMomentum alloc] init];
        targetMomentum.mass = pulserMomentum.mass;
        targetMomentum.velocity = (momentum.mass * momentum.velocity + pulserMomentum.mass * pulserMomentum.velocity)/pulserMomentum.mass;
        [self.pulser updateMomentum:targetMomentum];
        LNScrollViewMomentum *feedbackMomentum = [[LNScrollViewMomentum alloc] init];
        feedbackMomentum.mass = momentum.mass;
        feedbackMomentum.velocity = 0.f;
        return feedbackMomentum;
    }
}

@end
